### PR TITLE
[060] Improve style (Statement-first programming)

### PR DESCRIPTION
**Pull request by martinmoene**
And here's another. 

In this chapter there were a few (parenthetical) remarks that were hindering my reading and understanding of the sentence. I found the information not always important enough to devote words to it then and there, so I removed these. In other cases I tried to keep the main message together and added the remark as an extra sentence. You might judge differently ;)

I made the following change I'm uncertain of: *deployment machine*  => *development machine*

You write:
>There are several reasons and I will try to outline a few of them briefly.  
>
>**You do not know whether the Statement can ever be false until you see it evaluate as false**

To me it looks this is the only reason you describe (be it with three examples).

I've trouble to understand:  
- (Unless your Statement is evaluated prior to being fulfilled.)
-  Write false Statement (this step ensures that code has high testability)

At some point reading this chapter (can't remember where) I had the thought :
Have you come across the term [Future Perfect Thinking](http://c2.com/cgi/wiki?FuturePerfectThinking) ?

I made quite a few changes, but I've tried hard to "keep it your text". 

Let me know about any changes and clarifications you'd like to see.

cheers,
Martin

### CODE REVIEW

**Comment by martinmoene** [comment]
Some further explanation.

I removed the mention of (SCRUM) sprint commitment from:

>Another reason might be time pressure. Let's be honest -- we are all in a hurry, we are all under pressure and when this pressure is too high, it triggers heroic behaviors in us, especially when there is a risk of not making it with the sprint commitment.
>...
>If Specification is written at the end, it is often sacrificed on the altar of making it with the commitment, since the code is already written,

As 0) Not everybody is in a hurry, 1) not everyone does SCRUM, 2) it is not necessary for the subject of time pressure, and 3) most of all I suspect it may be problematic.

You may not intend to suggest that sprint-end-heroic behavior is normal/common or whatever, but I think mentioning it here is not supporting your main line --**time-pressure-induced** heroic behavior-- and may even harm it.

The scope of the work in a sprint is determined by estimating the time needed to implement its parts and see what likely fits. An estimate is not a guarantee and the commitment to the sprint does not mean a promise to complete it all [1][2]. (Other interesting reading on estimating [3][4].)

Not being able to complete all items of a sprint does not mean we must work harder, longer, differently to try and complete everything. However, the combination of the words *sprint commitment*, *time pressure* and *heroic behavior* do suggest that this is a common reaction and **leaves it at that**. As I don't think this location in your text is the right place and time to discuss it, it is perhaps best left out.

Hope this clarifies my approach.

cheers,
Martin

References:
[1] Serge Beaumont. ["Committing" to a Sprint and failing is a GOOD thing!](http://blog.xebia.com/2010/03/24/commiting-to-a-sprint-and-failing-is-a-good-thing/) March 24, 2010.
[2] Jose Luis Soria Teruel. [Commitment vs. Forecast: A subtle but important change to Scrum](https://www.scrum.org/About/All-Articles/articleType/ArticleView/articleId/95/Commitment-vs-Forecast-A-subtle-but-important-change-to-Scrum). October 01, 2011.
[3] Dan North. [The perils of estimation](http://dannorth.net/2009/07/01/the-perils-of-estimation/). July 1, 2009.
[4] Martin Fowler. [PurposeOfEstimation](http://martinfowler.com/bliki/PurposeOfEstimation.html). 27 February 2013.

**Comment by grzesiek-galezowski** [comment]
I wrote this chapter a long time ago. Looking at it now, I think it would benefit from some e-prime language ;-). I got an issue open (#38) to get back to this chapter and fix few things. I'll answer your concerns when I review all the changes, as I can't remember everything I wrote back then.

On sprint failures - I should have written that this was my observation that an end of a sprint looked like saving a sinking ship every time I looked (and commitment was being missed). This is backed by some other people's observation. I remember my intention was not to write that this is how things should be or that this is what SCRUM is about, so maybe I failed to state this intention. Speaking with some of my friends who work in several other companies, I came to think of it as being a common reaction. This is the reason I found it worth to discuss - because it was so common. I agree though that this has more to do with "missing a deadline" than with SCRUM in particular.

File: manuscript/060_Statement_First_Programming.md [hunk 1]
```diff
@@ -1,56 +1,56 @@
...
+Every accurate Statement fails when it isn’t fulfilled and passes when it is. That is one of the main reasons why we write it -- to receive this feedback. Also, after being fulfilled, the Statement becomes a part of the executable specification and starts failing as soon as the code stops fulfilling it, for example as the result of a mistake made during code rework. If you run a Statement for the first time only after it has been implemented and it evaluates to true, how do you know whether it really describes a need accurately? You did not ever watch it fail, so how do you know it ever will?
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: Could you also fix this one: "If you run a Statement for the first time only after it has been implemented and it evaluates to true". Looking at it now, I read it as if I meant implementing the code that makes the specification, not implementing what the specification describes.

maybe something like "If you run a Statement for the first time only after the behavior it describes has been implemented and it evaluates to true" ..?

File: manuscript/060_Statement_First_Programming.md [hunk 2]
```diff
@@ -63,7 +63,7 @@ public class CalculatorSpecification
...
+Now, imagine that you are writing this Statement post-factum as a unit test in an environment that has, let's say, more than thirty Statements. You have written the code and now you are just creating test after test “to ensure" the code works. Code, test -- pass, test -- pass, test -- pass. Usually you evaluate your code against the whole Specification, since it is easier than selecting what to evaluate each time. Besides you get more confidence this way that you don't make a mistake and break something that is already working. So, this is really: Code, Test -- all pass, test -- all pass, test -- all pass... Hopefully, you use some kind of snippets mechanism for creating new Statements, otherwise you might write something like this once in a while:
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: When the "(as you see, this is not my favorite reason for writing unit tests)" is left out, the  quotation marks around "to ensure" add no value in my mind. Let's remove them. By the way, now I see that I should have written this as a story with me as the main character, not the poor reader :-D.

File: manuscript/060_Statement_First_Programming.md [hunk 3]
```diff
@@ -78,15 +78,15 @@ public class CalculatorSpecification
...
+How does treating tests as Statements and evaluating them before making them true help here? **Because then, a Statement that starts off being evaluated as true is what DOES disturb your work flow.** In TDD, the work flow is: Statement -- unfulfilled -- fulfilled (and refactor, but that doesn't matter much for this discussion), Statement -- unfulfilled -- fulfilled, Statement -- unfulfilled -- fulfilled... So every time you miss the “unfulfilled" stage, you get feedback from your process that something suspicious is happening. This lets you investigate and fix the situation if necessary.
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: I'd change **Because then, a Statement that starts off being evaluated as true is what DOES disturb your work flow.** to just "Because then, a Statement that starts off being evaluated as true is what **does** disturb your work flow."

what was I thinking? :-D

File: manuscript/060_Statement_First_Programming.md [hunk 4]
```diff
@@ -110,13 +110,13 @@ public void ShouldRecognizeTimeSlotAboveMaximumAllowedAsInvalid()
 }
 ```
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: I'll finish reviewing this tomorrow. Thank you for your hard work. I have finishing another chapter on value objects in plans and then I'm getting back to this chapter to make it better. I agree with most of what I wrote, so I expect ca. 95% reuse, but as it is, it's written in an offensive "in-your-face" way that I thankfully unlearned some time ago :-).

File: manuscript/060_Statement_First_Programming.md (hunk 1 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: Yes, clear now (updated in PR).

File: manuscript/060_Statement_First_Programming.md (hunk 2 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: Indeed (updated in PR).

File: manuscript/060_Statement_First_Programming.md (hunk 3 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: ;) (updated in PR).

**Comment by martinmoene** [comment]
Meaning of Sprint commitment according to ref [1] above:
- fixing the scope of a Sprint.
- team pledges to each other that they will strive to reach the goal.
- *not* about accountability and holding a team in judgement after a Sprint.

File: manuscript/060_Statement_First_Programming.md [hunk 5]
```diff
@@ -110,13 +110,13 @@ public void ShouldRecognizeTimeSlotAboveMaximumAllowedAsInvalid()
...
 
-Note how the method `PerformForTimeSlotIn()`, which triggers the specified behavior is accidentally called BEFORE the mock is actually set up and the set up return value is never taken into account. By some strange coincidence, this error did not alter the expected end result so we did not even notice. It sometimes turns out like this, most often in case of various boundary values (nulls etc.).
+Note how the method `PerformForTimeSlotIn()`, which triggers the specified behavior, is accidentally called *before* the mock is actually set up and the value of `frame.GetTimeSlot_Returns` is not taken into account. Thus this erroneous value does not alter the expected end result, but we'll not notice it. It sometimes turns out like this, most often in case of various boundary values (nulls etc.).
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: I managed to take one more look today :-) 

I suggest to change "expected end result, but we'll not notice it" to something like "expected end result, and may pass unnoticed".

File: manuscript/060_Statement_First_Programming.md [hunk 6]
```diff
@@ -110,13 +110,13 @@ public void ShouldRecognizeTimeSlotAboveMaximumAllowedAsInvalid()
...
 
 #### 3. Using static data inside production code
 
-Once in a�while, you have to jump in and add some new Statements to some class Specification and some logic to the class itself. Let's assume that the class and its existing specification was written by someone else. Imagine this code is a�wrapper around your product XML configuration file. You decide to write your Statements AFTER applying the changes (�well", you can say, �I am all protected by the Specification that is already in place, so I�can make my change without risking regression, then just test my changes and it is all good...").
+Once in a while, you have to jump in and add some new Statements to an existing class Specification and some logic to the class itself. Let's assume that the class and its specification was written by someone else. Imagine this code is a wrapper around your product XML configuration file. You decide to write your Statements *after* applying the changes (“well", you say, “I am all protected by the Specification that is already in place, so I can make my change without risking regression, and then just test my changes and it is all good...").
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: Now that I look at it, "add some new Statements to an existing class Specification and some logic to the class itself" reads weird...

Maybe something like "add some new Statements to an existing Specification and some logic to the class it describes" would be better?

File: manuscript/060_Statement_First_Programming.md [hunk 7]
```diff
@@ -124,10 +124,9 @@ public class XmlConfigurationSpecification
   XmlConfiguration config = new XmlConfiguration(xmlFixtureString);
   
   //...
-  //...
 ```
 
-What it does is to set up an object used by every Statement. So, each Statement uses a�`config` object initialized with the same `xmlConfiguration` string value. The string is already pretty huge and messy, since it was made to contain what is required by all existing Statements. You need to write tests for is a�little corner case that does not need all this crap that is inside this string. So, you decide to start fresh and create a�separate object of `XmlConfiguration` class with your own, minimal string. Your Statement begins like this:
+What it does is to set up an object used by every Statement. So, each Statement uses a `config` object initialized with the same `xmlConfiguration` string value. The string is already pretty large and messy, since it contains all what is required by the existing Statements. You need to write tests for a little corner case that does not need all this crap inside this string. So, you decide to start afresh and create a separate object of the `XmlConfiguration` class with your own, minimal string. Your Statement begins like this:
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: "contains all **what** is required by the existing Statements" - would it be better to say "contains all **that** is required by the existing Statements"

File: manuscript/060_Statement_First_Programming.md [hunk 8]
```diff
@@ -135,39 +134,41 @@ var configuration = new XmlConfiguration(customFixture);
...
+What the...? Well, well, here is what happened: the author of this class applied a small optimization. He thought: “In this app, the configuration is only modified by members of the support staff and to do it, they have to shut down the system, so, there is no need to read the XML file every time an XmlConfiguration object is created. I can save some CPU cycles and I/O operations by reading it only once when the first object is created. Later objects will just use the same XML!". Good for him, not so good for you. Why? Because your custom XML string will never actually be used! (Unless your Statement is evaluated prior to being fulfilled.)
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: What would you say about:
"never actually be used! (Unless your Statement is evaluated prior to being fulfilled.)"
to 
"never actually be used (Unless your Statement is evaluated prior to being fulfilled)!"

File: manuscript/060_Statement_First_Programming.md (hunk 5 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: I slightly preferer "go unnoticed", although both seem correct. 

The only *reason* I can come up with is that the word 'pass' is already (strongly) connected to tests passing and failing...

File: manuscript/060_Statement_First_Programming.md [hunk 9]
```diff
@@ -135,39 +134,41 @@ var configuration = new XmlConfiguration(customFixture);
...
 ------------------------------------
 
-I will ask this question again: ever had to write a�requirement or design document for something that you already implemented? Was it fun? Was it valuable? Was it creative? No, I�do not think so. The same is with our executable specification. After we write the code, we have little motivation to specify what is already written -- some of the pieces of code �we can just see are correct", other pieces �we already saw working" when we copied our code over to our deployment machine and ran few sanity checks... The design is ready... Specification? Maybe next time...
+I will ask this question again: did you ever have to write a requirements or design document for something that you already implemented? Was it fun? Was it valuable? Was it creative? No, I do not think so. The same holds for our executable specification. After we've written the code, we have little motivation to specify what we wrote -- some of the pieces of code “we can just see are correct", other pieces “we already saw working" when we copied the code over to our development machine and ran a few sanity checks... The design is ready... Specification? Maybe next time...
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: Ok, I think I see what made you confused.

"other pieces “we already saw working" when we copied the code over to our deployment machine and ran a few sanity checks"

means:

"other pieces “we already saw working" when we compiled and deployed our changes and ran a few manual checks". I don't know how to express it better. This means building the product with the changes, deploying it on any environment (usually a test environment) and checking the logic by running few simple manual tests.

File: manuscript/060_Statement_First_Programming.md [hunk 10]
```diff
@@ -135,39 +134,41 @@ var configuration = new XmlConfiguration(customFixture);
...
+Another reason might be time pressure. When pressure becomes too high, it seems to trigger heroic behaviors in us. Then all of a sudden we drop all the “baggage", stop learning and experimenting, revert to all of the old “safe" behaviors and “save what we can!". If the Specification is written at the end, it is sacrificed, since the code already has been written, “and it will be tested anyway by real tests" like box tests, smoke tests and sanity tests. However, when starting with a Statement it is quite the opposite. Here a Statement evaluating to false is **a reason** to write any code. Thus, creating the Specification is an indispensable part of writing code.
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: I'm thinking of rewriting this paragraph a bit based on your corrections. Let me give it a try:

----------
Another reason might be time pressure and fear of deadlines. What I often see in teams I was in was a tendency to manifest behaviors that may be called "heroic", by which I mean being ready to sacrifice something "for greater good". This "greater good" was often seen in meeting the deadline. as for the sacrifices, it is often perceived that it's better to sacrifice something users won't notice rather than something they will. a Specification may then be pointed as such thing, especially when it written as the last thing, when production code is already written and, besides,  "it will be tested anyway by *real* tests" like black-box tests, smoke tests and sanity tests. 

Starting with a Statement brings a different point of view, where a false Statement is **a reason** to write any code. Thus, creating the Specification is an indispensable part of writing code.
------------


What do you think (it might need some corrections)?

File: manuscript/060_Statement_First_Programming.md (hunk 8 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: Not sure if the exclamation mark should be kept at such a distance! I agree that the parenthetical remark is part of the sentence with "will never actually be used". However these remarks hinder reading of the main line to some extend.

As I noted in the first remark of this PR, I've trouble to understand what
>Unless your Statement is evaluated prior to being fulfilled.

means: i.e. what exactly would make that the XML string be used?
As it's formulated now it's complicating what I thought I understood ;)

File: manuscript/060_Statement_First_Programming.md (hunk 9 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: This does make it clearer to me:
>"other pieces “we already saw working" when we compiled and deployed our changes and ran a few manual checks"

Updated in PR.

File: manuscript/060_Statement_First_Programming.md (hunk 7 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: that is probably the correct choice...

Updated in PR.

File: manuscript/060_Statement_First_Programming.md (hunk 6 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: Agree.

Updated PR.

File: manuscript/060_Statement_First_Programming.md [hunk 11]
```diff
@@ -135,39 +134,41 @@ var configuration = new XmlConfiguration(customFixture);
...
 
 1.  Write false Statement (this step ensures that code has high testability)
 2.  Write code to make the Statement true
 
-Now, how does it usually look like when we write the code first (extra steps marked with **strong text**):
+And here's how it often looks like when we write the code first (extra steps marked with **strong text**):
 
-1.  Write some production code (probably spans few classes until we are satisfied)
+1.  Write some production code (this probably spans a few classes until we are satisfied)
 2.  **Start writing unit tests**
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: Maybe: "Start writing *a* unit tests"?

File: manuscript/060_Statement_First_Programming.md [hunk 12]
```diff
@@ -135,39 +134,41 @@ var configuration = new XmlConfiguration(customFixture);
...
 3.  **Notice that unit testing the whole set of classes is cumbersome and unsustainable and contains high redundancy.**
 4.  **Restructure the code to be able to isolate objects and use mocks (this step ensures that code has high testability)**
 5.  Write unit tests
 
-What is the equivalent of the marked steps in Statement-first approach? Nothing! Doing these things is a�waste of time! Sadly, this is a�waste I�see done over and over again.
+What is the equivalent of the marked steps in the Statement-first approach? There is none! Doing these things is a waste of time! Sadly, this is a waste I encounter over and over again.
 
 Do you like wasting your time?
```
Comment (grzesiek-galezowski) [inline manuscript/060_Statement_First_Programming.md]: Phew, I'm finished with the review :-)

File: manuscript/060_Statement_First_Programming.md (hunk 11 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: Agree.

Updated in PR.

File: manuscript/060_Statement_First_Programming.md (hunk 10 above)
Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: Continuing  tomorrow...zzzz

Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: How about:

Another reason might be time pressure and fear of deadlines. Under these circumstances, the teams I worked in often showed a willingness to sacrifice something "for the greater good". Many times this "greater good" was seen in meeting the deadline. Rather sacrifice something users won't notice than sacrifice something they will might be the reasoning. It then becomes attractive to skip writing a Specification, especially when it is created after the production code has been written. Besides, "the code will be tested anyway by *real tests*" like black-box tests, smoke tests and sanity tests.

Starting with a Statement brings a different point of view, where a false Statement is **a reason** to write any code. Thus, creating the Specification is an indispensable part of writing code.

Comment (martinmoene) [inline manuscript/060_Statement_First_Programming.md]: Multiple: s/Another reason/Other reasons/

**Comment by grzesiek-galezowski** [comment]
Some of the answers I still owe you:

> "To me it looks this is the only reason you describe (be it with three examples)."

The reasons are:
 *  You do not know whether the Statement can ever be false until you see it evaluate as false
 * “Test-After" ends up as “Test-Never" 
 * "Test-After" leads to design rework

The structure is a bit messed up I made the issue #38 to address it after our last discussion on one of past pull requests (as far as I remember). Just give me some time :-) to finish the second chapter on value objects and I'll review the structure along with content to make it more logical.

> Because (unless your Statement is evaluated prior to being fulfilled), your custom xml string will never actually be used!

This is in fact confusing. I think I meant that the string is not used unless the Statement using it is the first one to initialize the static variable. Otherwise, the Statement might always pass or sometimes fail, but for the wrong reason. Then, what I think I forgot to mention is that by writing a Statement first and seing it pass somehow while it should not allows to detect the problem. I need to revisit this example a bit. For now, could you incorporate this feedback into the text?

> Write false Statement (this step ensures that code has high testability)

probably does not express what I meant as well. What I meant was that the place where we detect testability problems is moved to this step. Maybe it would be better to write something like:

" Write a Statement that is currently false (during this step, detect and correct testability issues even before the tested code is written)"

or something similar. What do you think?

**Comment by martinmoene** [comment]
Thanks, I'll have a look at it tomorrow.

**Comment by martinmoene** [comment]
I've changed the last two items above to:
>Good for him, not so good for you. Why? Because, due to how initialization works in C#, either the original XML string will be used for all Statements or your custom one! Thus the Statements in this Specification may pass or fail for the wrong reason. Writing a Statement first that you expect to fail and see it pass might help to detect this.

and:
>Write a Statement that is false to start with (during this step, detect and correct testability issues even before the tested code is written).

I'd like to suggest to defer more/other changes after merging this PR and continue working from the updated text.

cheers,
Martin

**Comment by grzesiek-galezowski** [comment]
Agree. I like the modifications you've made. Thank you for you hard work!
